src/hydro_system.hpp

```diff
@@ -583,6 +573,95 @@ void HydroSystem<problem_t>::FlattenShocks(amrex::MultiFab const &q_mf, amrex::M
 	});
 }
 
+// to ensure that physical quantities are within reasonable
+// floors and ceilings which can be set in the param file
+template <typename problem_t>
+void HydroSystem<problem_t>::EnforceLimits(amrex::Real const densityFloor, amrex::Real const pressureFloor, amrex::Real const speedCeiling,
+					   amrex::Real const tempCeiling, amrex::Real const tempFloor, amrex::MultiFab &state_mf)
+{
+
+	amrex::Real const rho_floor = densityFloor; // workaround nvcc bug
+	amrex::Real const P_floor = pressureFloor;
+	auto state = state_mf.arrays();
+
+	amrex::ParallelFor(state_mf, [=] AMREX_GPU_DEVICE(int bx, int i, int j, int k) noexcept {
+		amrex::Real const rho = state[bx](i, j, k, density_index);
+		amrex::Real const vx1 = state[bx](i, j, k, x1Momentum_index) / rho;
+		amrex::Real const vx2 = state[bx](i, j, k, x2Momentum_index) / rho;
+		amrex::Real const vx3 = state[bx](i, j, k, x3Momentum_index) / rho;
+		amrex::Real const vsq = (vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real const v_abs = std::sqrt(vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real Etot = state[bx](i, j, k, energy_index);
+		amrex::Real Eint = state[bx](i, j, k, internalEnergy_index);
+		amrex::Real Ekin = rho * vsq / 2.;
+		amrex::Real rho_new = rho;
+
+		if (rho < rho_floor) {
+			rho_new = rho_floor;
+			state[bx](i, j, k, density_index) = rho_new;
+			state[bx](i, j, k, internalEnergy_index) = Eint * rho_new / rho;
+			state[bx](i, j, k, energy_index) = rho_new * vsq / 2. + (Etot - Ekin);
+			if (nscalars_ > 0) {
+				for (int n = 0; n < nscalars_; ++n) {
+					state[bx](i, j, k, scalar0_index + n) *= rho / rho_new;
+				}
+			}
+		}
+
+		if (v_abs > speedCeiling) {
+			amrex::Real rescale_factor = speedCeiling / v_abs;
+			state[bx](i, j, k, x1Momentum_index) *= rescale_factor;
+			state[bx](i, j, k, x2Momentum_index) *= rescale_factor;
+			state[bx](i, j, k, x3Momentum_index) *= rescale_factor;
+		}
+
+		// Enforcing Limits on temperature estimated from Etot and Ekin
+		// re-obtain Ekin and Etot for putting limits on Temperature
+		Ekin = std::pow(state[bx](i, j, k, x1Momentum_index), 2.) / state[bx](i, j, k, density_index) / 2.;
```

Comment:
The other division-by-zero error occurs here. I'm not sure how to handle this case, since the kinetic energy is undefined. The temperature limits might have to be disabled if the density floor is zero.

Reply:
If the density is set to zero, energies should be set to zero as well. This will likely still cause crashes, but it is at least consistent.